codec/json.go

```diff
@@ -3,18 +3,23 @@ package codec
 import (
 	"bytes"
 
-	"github.com/cosmos/cosmos-sdk/codec/types"
-
 	"github.com/gogo/protobuf/jsonpb"
 	"github.com/gogo/protobuf/proto"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
 )
 
+var defaultJM = &jsonpb.Marshaler{OrigName: true, EmitDefaults: true, AnyResolver: nil}
+
 // ProtoMarshalJSON provides an auxiliary function to return Proto3 JSON encoded
 // bytes of a message.
 func ProtoMarshalJSON(msg proto.Message, resolver jsonpb.AnyResolver) ([]byte, error) {
 	// We use the OrigName because camel casing fields just doesn't make sense.
 	// EmitDefaults is also often the more expected behavior for CLI users
-	jm := &jsonpb.Marshaler{OrigName: true, EmitDefaults: true, AnyResolver: resolver}
+	jm := defaultJM
+	if resolver != nil {
+		jm = &jsonpb.Marshaler{OrigName: true, EmitDefaults: true, AnyResolver: resolver}
+	}
```

Comment:
what does this bring compared to the old (shorter) code?

Reply:
The default marshaler is cached. In many places we don't pass a resolver, so we can reuse a default one.

Reply:
Yes, you are right. `legacy.Cdc.UnmarshalBinaryLengthPrefixed` fails if we reduce the structure.
Thanks for catching. 

I've created an issue to trace it: https://github.com/cosmos/cosmos-sdk/issues/8323 (tentatively assigned to 0.42 milestone).